File: csharp/Domain/Models/RallyParticipant.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Domain.Models
{
    public class RallyParticipant : Identifiable<long>
    {
        public String FirstName { get; set; }
        public String LastName { get; set; }

        public Rally Rally { get; set; }
        public RallyTeam Team { get; set; }

        public RallyParticipant(Rally rally, RallyTeam team, String firstName, String lastName)
        {
            FirstName = firstName;
            LastName = lastName;
            Rally = rally;
            Team = team;
        }

        public override bool Equals(object? obj)
        {
            return obj is RallyParticipant participant &&
                   base.Equals(obj) &&
                   Id == participant.Id &&
                   FirstName == participant.FirstName &&
                   LastName == participant.LastName &&
                   Rally == participant.Rally &&
                   Team == participant.Team;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(base.GetHashCode(), Id, FirstName, LastName, Rally, Team);
        }
    }
}

File: java/domain/src/main/java/models/RallyParticipant.java
package models;

import java.util.Objects;

public class RallyParticipant extends Identifiable<Long> {
    private String firstName;
    private String lastName;
    private RallyTeam team;
    private Rally rally;

    public RallyParticipant(RallyTeam team, Rally rally, String firstName, String lastName) {
        this.team = team;
        this.rally = rally;
        this.firstName = firstName;
        this.lastName = lastName;
    }

    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public RallyTeam getTeam() {
        return team;
    }

    public Rally getRally() {
        return rally;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RallyParticipant)) return false;
        if (!super.equals(o)) return false;
        RallyParticipant that = (RallyParticipant) o;
        return Objects.equals(firstName, that.firstName) &&
                Objects.equals(lastName, that.lastName) &&
                Objects.equals(team, that.team) &&
                Objects.equals(rally, that.rally);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), firstName, lastName, team, rally);
    }
}
